server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -1159,6 +1159,32 @@ public <T> Set<T> readSet(Writeable.Reader<T> reader) throws IOException {
         return readCollection(reader, Sets::newHashSetWithExpectedSize, Collections.emptySet());
     }
 
+    /**
+     * Reads a set of objects. The set is expected to have been written using {@link StreamOutput#writeCollection(Collection)}} with
+     * a collection that contains no duplicates. The returned set is immutable.
+     *
+     * @return the set of objects
+     * @throws IOException if an I/O exception occurs reading the set
```

Comment:
Lets also document `IllegalArgumentException("duplicate element: " + e);` so it has more chances to be noticed here